src/transfer.jl

```diff
@@ -1,120 +1,211 @@
+        I[λ_ind] = 2π * all_mu_transfer_integral(τ_λ, view(S, :, λ_ind))
     end
     I
 end
 
 """
-    spherical_transfer(R, radii, α, S, μ_surface_grid)
+    spherical_transfer(α, S, τ_ref, α_ref, radii, μ_surface_grid)
 
 Perform radiative transfer along rays emerging at the μ values in `μ_surface_grid` in a spherically
-symmetric atmosphere with outer radius `R` [cm], and absorption coefficient, `α`, and source 
-function, `S`, resolved at radii `radii` [cm].
-"""
-function spherical_transfer(R, radii, α, S, μ_surface_grid)
-    r0 = radii[end] #radii of inner and outer atmosphere layers
-    
-    I_type = typeof(promote(radii[1], α[1], S[1], μ_surface_grid[1])[1])
-    I = Matrix{I_type}(undef, size(α, 2), length(μ_surface_grid))
-    
+symmetric atmosphere resolved at radii `radii` [cm]. See [`radiative_transfer`](@ref) for an 
+explantion of the arguments. Note that `radii` should be in decreasing order.
+
+Returns `(flux, intensity)`, where `flux` is the astrophysical flux, and `intensity`, a matrix of 
+shape (wavelengths × mu values), is the surface intensity as a function of μ.
+"""
+function spherical_transfer(α, S, τ_ref, α_ref, radii, μ_surface_grid)
+    R, r0 = radii[1], radii[end] #lower bound of atmosphere
+
+    #precompute for use in transfer integral
+    log_τ_ref = log.(τ_ref) 
+    τ_ref_α_ref = τ_ref ./ α_ref
+
+    #general element type with which to preallocate arrays
+    el_type = typeof(promote(radii[1], α[1], S[1], μ_surface_grid[1])[1])
+
+    #preallocations
+    #geometric path-length correction, and other wavelength-indenpendent integrand factors
+    integrand_factor = Matrix{el_type}(undef, size(α, 1), length(μ_surface_grid))
+    #the index of the lowest atmospheric layer pierced by each ray
+    lowest_layer_indices = Vector{Int}(undef, length(μ_surface_grid))
     for (μ_ind, μ_surface) in enumerate(μ_surface_grid)
         # impact parameter of ray
         b = R * sqrt(1 - μ_surface^2)
         
-        #calculate the index of the lowest layer the ray passes through
-        i = argmin(abs.(radii .- b))
+        #doing this with `findfirst` is messier at first and last index
+        i = argmin(abs.(radii .- b)) 
         if radii[i] < b
             i -= 1
         end
-        if i == 1 || i == 0
-            I[:, μ_ind] .= 0
-            continue
-        end #can eliminate for i == 1?
-        
-        #ds is the path length through each layer
-        ds = -diff(@. sqrt([R ; radii[1:i]]^2 - b^2)) 
-        dτ = α[1:i, :] .* ds
-        τ = cumsum(dτ, dims=1)
-        
-        for j in 1:size(τ, 2) #iterate over wavelength
-            I[j, μ_ind] = transfer_integral(τ[:, j], S[1:i, j]; plane_parallel=false)
+        lowest_layer_indices[μ_ind] = i
+
+        integrand_factor[1:i, μ_ind] = @. radii[1:i] ./ sqrt(radii[1:i]^2 - b^2) * τ_ref_α_ref[1:i]
+    end
+
+    #preallocations
+    I = Matrix{el_type}(undef, size(α, 2), length(μ_surface_grid)) #the surface intensity
```

Comment:
Sorry, I missed this before. You should also consider flipping the axes order on this matrix so that `μ_surface_grid` is along the fast-access-axis. This should make access in the following loop. It will also make the calculation of astrophysical flux faster.

Just a heads up, if you do this, you're going to need to modify line 111 to read `F = 2π * [Korg.trapezoid_rule(μ_surface_grid, μ_surface_grid .* I) for I in eachcol(I)]`